server/homebrew.api.js

```diff
@@ -33,19 +72,36 @@ const MAX_TITLE_LENGTH = 100;
 
 const getGoodBrewTitle = (text)=>{
 	const tokens = Markdown.marked.lexer(text);
- 	return (tokens.find((token)=>token.type == 'heading' ||	token.type == 'paragraph')?.text || 'No Title')
+ 	return (tokens.find((token)=>token.type === 'heading' || token.type === 'paragraph')?.text || 'No Title')
 				 .slice(0, MAX_TITLE_LENGTH);
 };
 
 const excludePropsFromUpdate = (brew)=>{
 	// Remove undesired properties
-	const propsToExclude = ['views', 'lastViewed'];
+	const propsToExclude = ['_id', 'views', 'lastViewed', 'editId', 'shareId', 'googleId'];
 	for (const prop of propsToExclude) {
 		delete brew[prop];
 	}
 	return brew;
 };
 
+const excludeStubProps = (brew)=>{
+	const propsToExclude = ['text', 'textBin', 'renderer', 'pageCount', 'version'];
+	for (const prop of propsToExclude) {
+		brew[prop] = undefined;
```

Comment:
How does `= undefined` differ from `delete` we use elsewhere?

Reply:
I found that `delete` doesn't work on Mongoose object instances, whereas ` = undefined` did.

Reply:
Perhaps we can do the `to object` conversion right after retrieving a Mongo brew in `getBrew` so we can be sure the result of `getBrew` is always normal Javascript object we can manipulate? Instead of carrying around an unknown object type that we have to keep cloning or checking the type?

Then in the case we actually need a mongo object to "save", etc., we just do a simple `mongo.find(editId)` or whatever right where we need it.

Reply:
I'm not _opposed_ to that plan, but I will say that the only case in which it returns a mongoose object is when it's fetched for editing. In all other cases it does the `toObject` call. I feel like the switch to setting values to undefined is small enough that the benefit of not having to make a whole additional call to the database to end up updating it is worth it. 

Happy to impl your suggestion though if it seems problematic enough though

Reply:
Not just for the delete thing. I think it will be helpful to guarunteed the object type coming out of that function. Especially in the future if we make more changes, I don't want to leave a trap behind for a future developer.

Reply:
I think that's basically what we had before anyway. Rather than preaching each action with getBrew, we just called mongo.find inline when needed.

Reply:
TLDR; `getBrew` gets the brew data to send to the user (maybe should be `getBrewForUser` or something). If we just need a mongo object so we can save, we don't need the whole process of "retrieve the brew, unzip the compressed brew text from `textBin`, get the brew from Google Drive, etc. This also means we can do similar to #2154 in the `find` by adding a "fields" parameter so the mongo object we retrieve for saving is only the `_id` or something, which should greatly reduce the amount of database/server bandwidth for large brews, instead of grabbing the entire 5MB document every time the autosave triggers, unzipping it, and then zipping the new version and sending a whole new 5MB back to the database.

Reply:
So this would be the workflow if I'm understanding correctly:
- Retrieves the brew during `getBrew` and converts it to a standard object
- Updates the object and deletes values where applicable for the current iteration
- Fetch the Mongo object
- `Object.assign(stub, updated)` or a `_.merge` call of some kind to join the values from the updated object to the Mongo object

In most cases that works fine, but when transferring to Google it will be deleting the text and textBin from the updated object, meaning the object won't have those values, so the assign/merge will not change the values on the Mongo stub. 

For that case we would then need to run `brew.<x> = undefined` where x is the GDrive stored props, which just brings us back to the method as it is now.

Reply:
The flow is basically what I was thinking, yes.

Ok, we can leave `brew.x = undefined` for managing the stubs in this special case then. However, the other parts (`getBrew` returns a standard object, and for saving we fetch the Mongo object with the minimal number of fields to reduce data)

Reply:
How would you recommend applying the values from the standard object to the Mongoose object?

Reply:
So I know as soon as we start working with array types more `merge` will be a problem -- it merges arrays recursively, which won't work for tags and authors. The same issue applies for undefined, for example,
```js
const x = _.merge({ tags: ['tag1', 'tag2'] }, { tags: ['tag1'] })
x // { a: ['tag1', 'tag2'] }

const y = _.merge({ a: 'test' }, { a: undefined })
y // { a: 'test' }
```
I'll try the spread operator today and see if it copies the Mongoose methods.
```js
const z = { ...{a: 'test'}, ...{a: undefined} }
z // { a: undefined }
```

Reply:
If not `_.merge` there should be some similar approach.

I guess we don't need to merge properties as much as overwrite them, no? Does `_.assign` do what we need? By the time the JS object gets to this point, can we assume it looks exactly how the final saved object should look, minus ids and stuff? So if we previously had `tags: [tag1, tag2, tag99]` and now the brew has `tags: [tag2, tag300]` we would want to just fully overwrite `tags` to be the new array, instead of merging the arrays.

Reply:
Sorry I've been away for a bit, I tried to make this work but both using `new HomebrewModel(brew)` and `_.assign(await HomebrewModel.find({ _id: req.brew.id }), brew)` the HomebrewModel gets _very_ upset and errors out saying it's a duplicate key on `_id`. I don't think this is going to be a viable change without some sneaky js work, and I'd much prefer no sneaky js work and `toObject` called in a few extra places. Thoughts?

Reply:
Never mind! I was not performing the find correctly. 😬

Reply:
Can you point out which lines are causing the conflict?

But as a guess, I'm thinking in the singular `toObject` we could clear the `_id` parameter from the resulting JSObject (its only useful as a unique identifier for Mongo. We dont really use it), and when we need to look up a document in the database, we just use the editID or shareID. No need to try preserving the original mongo ID between transfers, etc. Just our edit/share IDs.

Reply:
Lol. Alright. Ask away if you run into something else.